### ISSUE TITLE
Sublanguages - for example javascript in html

### ISSUE THREAD

**Issue by LaKing** [isagalaev/highlight.js#877]
I noticed that in recent highlight.js versions sublanguages are not detected correctly.
An example.
```html
<script type="text/javascript">
    var abc = require('abc');
</script>
```

The parsed highlighted HTML is:

```html
<span class="hljs-tag">&lt;<span class="hljs-title">script</span> <span class="hljs-attribute">type</span>=<span class="hljs-value">"text/javascript"</span>&gt;</span>
<span class="coffeescript">
    <span class="hljs-reserved">var</span> abc = <span class="hljs-built_in">require</span>(<span class="hljs-string">'abc'</span>);
</span>
<span class="hljs-tag">&lt;/<span class="hljs-title">script</span>&gt;</span>
```
.. It seems to detect coffescript. As far as I see the script type is not parsed from the code.
IMO javascript should be the default, and type should enforce other languages. I'm not a fan of language detection techniques, it should be the last option.

Furthermore, when using the API, and processing strings line by line hljs.highlight passes XML when on XML name input - correctly - but then puts an "undefined" tag at the sublanguage start, and in the next line, instead of passing this information thru the stack, it starts language detection, and brings up all possible languages, .. then the closing tag again is undefined.

To keep the example consistent:
```html
<div class="" id="magicdomid2">
<span class="hljs-tag">&lt;<span class="hljs-title">script</span> <span class="hljs-attribute">type</span>=<span class="hljs-value">"text/javascript"</span>&gt;</span>
<span class="undefined"></span>
</div>
<div class="" id="magicdomid3">
<span class="coffeescript">&nbsp;&nbsp;&nbsp; <span class="hljs-reserved">var</span> abc = <span class="hljs-built_in">require</span>(<span class="hljs-string">'abc'</span>);</span>
</div>
<div class="" id="magicdomid4">
<span class="undefined"></span>
<span class="hljs-tag">&lt;/<span class="hljs-title">script</span>&gt;</span></div>
```

As far as I remember this worked correctly in previous versions. ...

**Comment by isagalaev** [comment]
> .. It seems to detect coffescript. As far as I see the script type is not parsed from the code.
IMO javascript should be the default, and type should enforce other languages. I'm not a fan of language detection techniques, it should be the last option.

To put this problem in context, we used to highlight the contents of the `<script...>` with JavaScript only. It was changed to automatic detection due to increasing popularity of other languages.

Unfortunately with the current parser we can't neither pick the hint from the type attribute nor treat a language preferentially. The former would require a pretty non-trivial HTML-specific special case in the language-agnostic core, and for the latter to work we would need to define when to choose other languages beside the default one.

In other words, we know that it currently sucks but it's the best we can do right now.

**Comment by isagalaev** [comment]
I fixed this particular example in 558973a2.

**Comment by LaKing** [comment]
I updated to the latest git version, and the issues still exist with the API.
- undefined span class when a sublanguage starts
- all sorts of random languages detected in the next lines.

When using highlighting for a whole document, the detection is more likely to be accurate, as there is a bigger codebase to work with. But with the API call, processing line-by line detection is IMO not really possible. 

Should programmers using the API care for sublanguage detections? In that case the API might need an extension, like a sublanguage level parameter passed on and on similar to the stack? 
.. as in this case, I will have to tell the highlight.js api call, what sublanguage it should use.

**Comment by LaKing** [comment]
Here another example:
```html
<script type="text/javascript">
    //var abc = require('abc');
</script>
```

is parsed to:
```html
<code class="xml hljs">
<span class="hljs-tag">&lt;<span class="hljs-title">script</span> <span class="hljs-attribute">type</span>=<span class="hljs-value">"text/javascript"</span>&gt;</span>
<span class="cal">
    //<span class="hljs-keyword">var</span> abc = require(<span class="hljs-string">'abc'</span>);
</span><span class="hljs-tag">&lt;/<span class="hljs-title">script</span>&gt;
</span>
</code>
```

As you see, instead of javascript, with the // startinga  comment "cal" is detected.

**Comment by LaKing** [comment]
> To put this problem in context, we used to highlight the contents of the <script...> with JavaScript only. It was changed to automatic detection due to increasing popularity of other languages.

I see. My problem is that backward compatibility broke with this, ...

The proper solution IMO would be to keep javascript as default, and eventually add an optional parameter that overwrites this default. Once html could be parsed that language detection is less necessery, but still has the right to exists. Also, according to Wc3 recommendations 
```<META http-equiv="Content-Script-Type" content="type">``` 
.. should define the default language, but then again this is parsing html.

>In other words, we know that it currently sucks but it's the best we can do right now.

I still love it, but it would be bad to stuck at the version I used to use.

**Comment by isagalaev** [comment]
I think I've found a better solution than just reverting it to hard-coded JavaScript. We could extend the `subLanguage` syntax to support several languages and run auto-detect among them. There are probably only a few of those that realistically exist in the wild within `<script>..</script>` and this way we eliminate weird detection choices that can't exist in practice.

(As an aside, I think we can remove the `subLanguageMode` attribute and make specified sub-languages to be always continuous.)

**Comment by Sannis** [comment]
Totally agreed)

**Comment by isagalaev** [comment]
BTW, I could only find one language that was ever requested beside JavaScript — Handlebars. Any other suggestions, guys?

**Comment by isagalaev** [comment]
… oh, and ActionScript too (we've even got that one in the test suit!)

**Comment by LaKing** [comment]
Actually other scripting languages didnt make it. I guess 99% of all cases, script is starting javascript.

When W3C was working on HTML5, they discovered all browsers have "text/javascript" as the default script type, so they standardized it to be the default value. Hence, type is not needed within the script tag when using javascript. The language attribute has been deprecated ...

http://www.w3.org/TR/html5/scripting-1.html#attr-script-type

**Comment by isagalaev** [comment]
@LaKing there's no argument that JavaScript is the default value (and by far most popular). As I said earlier, we simply cannot support this in our generalized parser.

Anyway, I've just pushed the fix that constrains possible HTML scripting languages to just three and I believe it should take care of any practical issues.
